frappe/erpnext
# feat: Improve Product Bundle handling

## 1. Gross Profit Report

_Problem:_
- When grouped by Invoice, the Gross Profit Report isn't very readable, with several rows being displayed for each Invoice, representing each Item in the Invoice.
- In the case of Product Bundles, only the Parent Item is displayed, and not the Child Items.

_Proposed Solution:_
- Display the data in a tree format, with the Sales Invoice at the first level, Product Bundles and non-Bundle Items at the second level, and Child Items at the third.

<details>
<summary>Images</summary>

_Before:_

![Screenshot 2021-08-25 at 3 10 01 AM](https://user-images.githubusercontent.com/25903035/130693689-720c130b-9b51-43e0-a5c2-579e21c403e7.png)

_After:_

![Screenshot 2021-08-25 at 12 35 18 AM](https://user-images.githubusercontent.com/25903035/130674996-a6d891e4-4220-4574-87fc-5b32bca6992a.png)
</details>

## 2. Selling Settings

_Proposed Changes:_

- Add a new checkbox: _Calculate Product Bundle Price based on Child Items' Rates_

![Screenshot 2021-08-24 at 11 46 10 PM](https://user-images.githubusercontent.com/25903035/130668773-e647e86a-acf3-4365-884c-2de81db7eb10.png)

On enabling _Calculate Product Bundle Price based on Child Items' Rates_: 
- The Rate column of Packed Items will be made editable.
- The rate and price of Product Bundles in the Items table will be updated based on the rates of its child Items.

_Note_:
If the Rate of the Product Bundle is changed now, it will get reset to the sum based on the rates of its child items on saving the doc.

<details>
<summary>Example</summary>

_Bundle Item Rates:_

![Screenshot 2021-08-24 at 11 54 33 PM](https://user-images.githubusercontent.com/25903035/130669895-31bfc79b-1eb3-4a85-b662-3094a6590942.png)

_Parent Item's Rate and Amount:_

![Screenshot 2021-08-24 at 11 52 23 PM](https://user-images.githubusercontent.com/25903035/130669623-08bcfd29-f2bd-40c4-bb2b-51c99eb2e348.png)

</details>

## 3. Quotation

_Proposed Change:_

- Add a new table called _Bundle Items_ to display the child Items of all Parent Bundles listed in the Items table.

![Screenshot 2021-08-27 at 2 25 16 AM](https://user-images.githubusercontent.com/25903035/131034962-adb013fe-1210-4866-b15e-24cf7f9f789a.png)

## 4. Purchase Order

- When creating a Purchase Order from a Sales Order, if there are any Product Bundles in the Items table, they'll be replaced with their child Items.

<details>
<summary>Images</summary>

_Sales Order:_

![Screenshot 2021-08-30 at 9 18 27 PM](https://user-images.githubusercontent.com/25903035/131527525-f18212bf-4470-478f-980a-ad2055694671.png)


_Purchase Order:_

![Screenshot 2021-08-31 at 8 35 47 PM](https://user-images.githubusercontent.com/25903035/131527596-506c7f1f-0f55-40f9-bea0-2ecc29216bcd.png)

</details>

- Add a new read-only field called Product Bundle in the Items table to store the name of the Product Bundle when the mapping specified above takes place.

![Screenshot 2021-08-31 at 8 38 30 PM](https://user-images.githubusercontent.com/25903035/131528101-ae0d98c7-9067-40fe-8140-af504b2d38fa.png)

- Make Sales Order field editable so POs can be linked with them later.

![Editable SO Link](https://user-images.githubusercontent.com/25903035/131030804-b93bc3e1-c01e-4d5b-aa45-e550e2e73db6.png)

## 5. Purchase Receipt

- Add a new read-only field called Product Bundle in the Items table to store the names of Product Bundles when a Purchase Receipt is created from a Purchase Order that was created from a Sales Order containing Product Bundles.

![Screenshot 2021-09-01 at 4 15 09 PM](https://user-images.githubusercontent.com/25903035/131659002-9997c664-d7e3-415b-97ec-3da4f362e32b.png)

## 6. Purchase Invoice

- Add a new read-only field called Product Bundle in the Items table to store the names of Product Bundles when a Purchase Invoice is created directly from a Purchase Order that was created from a Sales Order containing Product Bundles, or from a Purchase Receipt that was created from such a Purchase Order.

![Screenshot 2021-09-01 at 4 14 17 PM](https://user-images.githubusercontent.com/25903035/131659026-901355f4-8b09-4730-bae4-459eace28683.png)

`no-docs`

<details>
<summary>Testing Info</summary>

_For Gross Profit Report:_
1. Open the report.
2. Set 'Group By' to 'Invoice'.
3. Verify if the items are displayed in a tree format with the Invoice at the top, followed by Product Bundles/Non-Bundle Items, and then the Child Items(if any).
(If confused, refer image under Gross Profit Report > Images > After of this PR description)


_For Calculate Product Bundle Price based on Child Items' Rates_:
1. Open Selling Settings.
2. Check _Calculate Product Bundle Price based on Child Items' Rates_.
3. Create a Product Bundle.
4. Create a new Quotation.
5. Add the Product Bundle to the Items table of the Quotation.
6. Save.
7. Check the Bundle Items table. The sum of (Qty x Rate) of all the Bundle Items with parent as the Product Bundle in question should be equal to the new Amount of the Product Bundle in the Items table, and its rate should be this new Amount/Qty.
8. Also check if the Rate column of the Bundle Items table is now editable.
9. Open Selling Settings again and uncheck _Calculate Product Bundle Price based on Child Items' Rates_.
10. Now check if the Rate column of the Bundle Items table is read-only.
11. If you change the rate of the Product Bundle in the Items table now, it will not be reset to the Amount based on the Bundle Items' rates on saving anymore.

_For Quotation:_
1. Create a Product Bundle.
2. Create a new Quotation.
3. Add the Product Bundle to the Items table of the Quotation.
4. Save.
5. Check if the Bundle Items table was created.

_For Purchase Order:_
1. Create a new Purchase Order.
2. Add an Item.
3. Expand the Item row.
4. Check if the Sales Order field under the 'Warehouses and References' section is editable.

1. Create a new Sales Order.
2. Now try adding a Product Bundle, save and submit.
3. Create > Purchase Order.
4. Check if the Packed Items in the SO had replaced the Product Bundle in the Items table of the PO.

</details>
